### PR TITLE
[core] split validate and assign defaults and testing

### PR DESCRIPTION
**Pull request by armingeiser**
Splits the implementation of the `ValidateAndAssignDefaults` into `AssignDefaults` and `ValidateDefaults`. the same for the recursive versions.

This allows more flexible handling of special cases of default parameters.

E.g. if some defaults should be assigned to a parameter object, without checking all parameters which are already there.

### CODE REVIEW

**Comment by armingeiser** [comment]
By the way: are the versions with non const function argument `(Parameters& rDefaultParameters)` really needed? I think the `rDefaultParameters` should never be modified in those functions...

**Review by philbucher — COMMENTED** [review]
+1 for only having the function that accepts a `const` argument in case it should not be modified

Will take a closer look later

**Comment by RiccardoRossi** [comment]
guys 2 things:

1 - you should not pass by reference a parameters object because of how it is implemented
2 - the non const version is needed as the modifications of the AssignDefault need to be stored in the parameters...

**Comment by loumalouomega** [comment]
Even with the new implementation?

**Comment by armingeiser** [comment]
@RiccardoRossi 

1. The currently implemented functions in the kratos_parameter.h e.g. https://github.com/KratosMultiphysics/Kratos/blob/ea96ef46647d11c868eb35c17cab775366166e5c/kratos/includes/kratos_parameters.h#L734
already do that. So this should be changed too, right? 

2. The modifications should only be applied to the object where the method is called, not to the argument which are the defaults. So the method is non const, but the argument could be const.

**Comment by RiccardoRossi** [comment]
@loumalouomega in this the behaviour should not be changed. ProjectParameters is a handle to a library below that.

regarding the const, and thinking about, we are not modifying the default parameters, so it can indeed be const

**Comment by RiccardoRossi** [comment]
(point 2, implies that my previous comment was wrong...we can leave only the const version, and in this context most probably the & is ok)

**Comment by armingeiser** [comment]
I removed the overloads with a non const argument. Tests are running.

File: kratos/includes/kratos_parameters.h [hunk 1]
```diff
@@ -726,36 +726,53 @@ class KRATOS_API(KRATOS_CORE) Parameters
     bool HasSameKeysAndTypeOfValuesAs(Parameters& rParameters);
 
     /**
-     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the rDefaultParameters.
+     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the rDefaultParameters. (const version)
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the rDefaultParameters.
```

File: kratos/includes/kratos_parameters.h [hunk 2]
```diff
@@ -726,36 +726,53 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
      * This version of the function only walks one level, without descending in the branches
      * @param rDefaultParameters Parameters of reference which we use to check
      */
-    void ValidateAndAssignDefaults(const Parameters& rDefaultParameters);
+    void AssignDefaults(const Parameters& rDefaultParameters);
 
     /**
-     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the defaults.
-     * @details If the parameters contain values that do not appear in the defaults, an error is thrown, whereas if a parameter is found in the defaults but not in the Parameters been tested, it is copied to the parameters.
+     * @brief This function is designed to verify that the parameters under testing contain at least all parameters prescribed by the rDefaultParameters. (const version)
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
     * @brief This function is designed to verify that the parameters under testing contain at least all parameters prescribed by the rDefaultParameters.
```

File: kratos/includes/kratos_parameters.h [hunk 3]
```diff
@@ -726,36 +726,53 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
-     * @details If the parameters contain values that do not appear in the defaults, an error is thrown, whereas if a parameter is found in the defaults but not in the Parameters been tested, it is copied to the parameters.
+     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters. (const version)
+     * @details If the parameters contain values that do not appear in the defaults, an error is thrown.
+     * This version of the function only walks one level, without descending in the branches
+     * @param rDefaultParameters Parameters of reference which we use to check
+     */
+    void ValidateDefaults(const Parameters& rDefaultParameters);
+
+    /**
+     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters. (const version)
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters.
```

File: kratos/includes/kratos_parameters.h [hunk 4]
```diff
@@ -726,36 +726,53 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
      * This version walks and validates the entire json tree below the point at which the function is called
      * @param rDefaultParameters Parameters of reference which we use to check
      */
-    void RecursivelyValidateAndAssignDefaults(Parameters& rDefaultParameters);
+    void RecursivelyAssignDefaults(const Parameters& rDefaultParameters);
 
     /**
-     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the defaults. (const version)
-     * @details If the parameters contain values that do not appear in the defaults, an error is thrown, whereas if a parameter is found in the defaults but not in the Parameters been tested, it is copied to the parameters.
+     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters. (const version)
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters.
```

File: kratos/includes/kratos_parameters.h [hunk 5]
```diff
@@ -726,36 +726,53 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
      * This version of the function only walks one level, without descending in the branches
      * @param rDefaultParameters Parameters of reference which we use to check
      */
-    void ValidateAndAssignDefaults(Parameters& rDefaultParameters);
+    void ValidateAndAssignDefaults(const Parameters& rDefaultParameters);
 
     /**
-     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the rDefaultParameters. (const version)
-     * @details If the parameters contain values that do not appear in the rDefaultParameters, an error is thrown, whereas if a parameter is found in the rDefaultParameters but not in the Parameters been tested, it is copied to the parameters.
+     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the defaults. (const version)
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the defaults.
```

File: kratos/includes/kratos_parameters.h [hunk 6]
```diff
@@ -726,36 +726,53 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
+     * @details If the parameters contain values that do not appear in the defaults, an error is thrown, whereas if a parameter is found in the defaults but not in the Parameters been tested, it is copied to the parameters.
+     * This version walks and validates the entire json tree below the point at which the function is called
+     * @param rDefaultParameters Parameters of reference which we use to check
+     */
+    void RecursivelyValidateAndAssignDefaults(const Parameters& rDefaultParameters);
+
+    /**
+     * @brief This function is designed to verify that the parameters under testing contain at least all parameters prescribed by the rDefaultParameters. (const version)
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
     * @brief This function is designed to verify that the parameters under testing contain at least all parameters prescribed by the rDefaultParameters.
```

**Review by philbucher — COMMENTED** [review]
some minor comments, haven't looked at it in detail yet

File: kratos/includes/kratos_parameters.h [hunk 7]
```diff
@@ -731,31 +731,48 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
 
     /**
-     * @brief This function is designed to verify that the parameters under testing match the form prescribed by the defaults. (const version)
-     * @details If the parameters contain values that do not appear in the defaults, an error is thrown, whereas if a parameter is found in the defaults but not in the Parameters been tested, it is copied to the parameters.
+     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters.
+     * @details If the parameters contain values that do not appear in the defaults, an error is thrown.
+     * This version of the function only walks one level, without descending in the branches
+     * @param rDefaultParameters Parameters of reference which we use to check
+     */
+    void ValidateDefaults(const Parameters& rDefaultParameters);
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]: this should be a const function

File: kratos/includes/kratos_parameters.h [hunk 8]
```diff
@@ -731,31 +731,48 @@ class KRATOS_API(KRATOS_CORE) Parameters
...
+
+    /**
+     * @brief This function is designed to verify that the all parameteters in the parameters under testing have a matching parameter prescribed by the rDefaultParameters.
+     * @details If the parameters contain values that do not appear in the defaults, an error is thrown.
      * This version walks and validates the entire json tree below the point at which the function is called
      * @param rDefaultParameters Parameters of reference which we use to check
      */
-    void RecursivelyValidateAndAssignDefaults(const Parameters& rDefaultParameters);
+    void RecursivelyValidateDefaults(const Parameters& rDefaultParameters);
```
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]: same

**Review by philbucher — CHANGES_REQUESTED** [review]
why no tests for `RecursivelyValidateDefaults`?

**Comment by philbucher** [comment]
other than that it looks good to me, thanks for refactoring, I think this will be really useful

**Comment by armingeiser** [comment]
fixed it

**Comment by pooyan-dadvand** [comment]
Is this backward compatible?

**Comment by armingeiser** [comment]
Yes, the old function that does both, validating and assigning defaults at the same time, is still there. Only internally the function has been split up and now the two parts can also be called individually.

**Comment by dbaumgaertner** [comment]
Just to add that this is a very welcome change, as we were struggling with this missing flexibility quite some times. Thanks @armingeiser for the effort!

**Comment by roigcarlo** [comment]
For the record, this is tightly related related with #2241. Linking for future references.

**Comment by RiccardoRossi** [comment]
the assigndefaults is actually equivalent to merging the defaultparams into the settings no?

if so i agree this is a very welcome contribution

**Review by philbucher — DISMISSED** [review]
removing my blocker
please wait until @KratosMultiphysics/technical-committee approves

**Comment by armingeiser** [comment]
@RiccardoRossi Yes, AssignDefaults merges the missing settings into the parameters object.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee agrees with the intention of this PR and see it necessary. However would like to clarify more the name of `AssignDefaults` with something like `AddMissingParameters`. Any suggestion?

**Comment by armingeiser** [comment]
I think `AddMissingParameters` is already quite descriptive. It is better than `MergeParameters`, because the function does not overwrite already existing settings (if i think about merging in context of e.g. Git, thats what i would expect from a merge function).

**Comment by armingeiser** [comment]
I renamed the function `AssignDefaults` to `AddMissingParameters`. It is ready for a final review from my point of view.

**Comment by philbucher** [comment]
ping, I believe this is being accepted and blocks many developments / would make many developments easier :)

**Comment by ddiezrod** [comment]
Is this ready to be merged? It would help us a lot

**Comment by armingeiser** [comment]
@ddiezrod the implementation is ready, just someone from @KratosMultiphysics/technical-committee has to approve it.

**Review by RiccardoRossi — APPROVED** [review]
i think it is ok now!

thx!

**Pull request merged**